File: api/service/nap.py
import asyncio
import re

import arrow
import bs4
import pydantic

import api.models.nap.content
import api.models.nap.list
from api.utils import request_async
from api.utils.timezone import get_tzinfo


async def get_ann_list():
    response_data = await request_async(
        "https://announcement-static.mihoyo.com/common/nap_cn/announcement/api/getAnnList",
        {
            "game": "nap",
            "game_biz": "nap_cn",
            "lang": "zh-cn",
            "bundle_id": "nap_cn",
            "platform": "pc",
            "region": "prod_gf_cn",
            "level": 40,
            "channel_id": 1,
            "uid": 10000000,
        },
    )

    return api.models.nap.list.Model.model_validate(response_data)


async def get_ann_content():
    response_data = await request_async(
        "https://announcement-static.mihoyo.com/common/nap_cn/announcement/api/getAnnContent",
        {
            "game": "nap",
            "game_biz": "nap_cn",
            "lang": "zh-cn",
            "bundle_id": "nap_cn",
            "platform": "pc",
            "region": "prod_gf_cn",
            "level": 40,
            "channel_id": 1,
            "uid": 10000000,
        },
    )
    return api.models.nap.content.Model.model_validate(response_data)


class NapGachaInfo(pydantic.BaseModel):
    ann_id: int
    title: str
    image: str
    content: str
    start_time: str | None
    end_time: str | None


class NapProgress(pydantic.BaseModel):
    text: str | None
    percent: float | None


class NapResponse(pydantic.BaseModel):
    progress: NapProgress
    gacha_info: list[NapGachaInfo]


async def get_nap_gacha_info():
    ann_list, ann_content = await asyncio.gather(get_ann_list(), get_ann_content())
    version_info = ann_list.get_version_info()
    timezone = ann_list.data.timezone
    progress_percent = None
    progress_text = None
    gacha_info: list[NapGachaInfo] = []

    if version_info:
        start_time = arrow.get(version_info.start_time, get_tzinfo(timezone))
        end_time = arrow.get(version_info.end_time, get_tzinfo(timezone))
        current_time = arrow.now("Asia/Shanghai")
        if start_time <= current_time <= end_time:
            progress_percent = (current_time - start_time) / (end_time - start_time)
            end_time_humanize = end_time.humanize(locale="zh", granularity=["day", "hour", "minute"])
            progress_text = f"{start_time:YYYY-MM-DD HH:mm:ss} ~ {end_time:YYYY-MM-DD HH:mm:ss} （{end_time_humanize}结束）"

    for i in ann_content.get_gacha_info():
        # 常驻频段
        stable_channel = "「热门卡司」调频说明"
        # 邦布频段
        bangboo_channel = "「卓越搭档」调频说明"
        # 永久频段
        permanent_channel_list = [
            stable_channel,
            bangboo_channel,
        ]
        if i.subtitle in permanent_channel_list:
            continue

        content_soup = bs4.BeautifulSoup(i.content, "html.parser")
        info_table_time = content_soup.select_one("table tr:nth-child(2) td")
        text = info_table_time.text if info_table_time else ""
        t = re.search(
            r"(?:(.*?后)|(.*?)（服务器时间）)~(.*?)（服务器时间）",
            text,
            re.MULTILINE,
        )
        start_time = None
        end_time = None
        match t.groups() if t else None:
            case [start_str, None, end_time]:
                end_time = arrow.get(end_time, tzinfo=get_tzinfo(timezone))
                end_time_humanize = end_time.humanize(locale="zh", granularity=["day", "hour", "minute"])
                start_time = f"{start_str}"
                end_time = f"{end_time:YYYY-MM-DD HH:mm:ss} （{end_time_humanize}）"
            case [None, start_time, end_time]:
                start_time = arrow.get(start_time, tzinfo=get_tzinfo(timezone))
                end_time = arrow.get(end_time, tzinfo=get_tzinfo(timezone))
                start_time_humanize = start_time.humanize(locale="zh", granularity=["day", "hour", "minute"])
                end_time_humanize = end_time.humanize(locale="zh", granularity=["day", "hour", "minute"])
                start_time = f"{start_time:YYYY-MM-DD HH:mm:ss} （{start_time_humanize}）"
                end_time = f"{end_time:YYYY-MM-DD HH:mm:ss} （{end_time_humanize}）"

            case _:
                pass

        gacha_info.append(
            NapGachaInfo(
                ann_id=i.ann_id,
                title=i.subtitle,
                image=i.banner,
                content=i.content,
                start_time=start_time,
                end_time=end_time,
            )
        )

    return NapResponse(
        progress=NapProgress(percent=progress_percent, text=progress_text),
        gacha_info=gacha_info,
    )

File: src/service/nap.ts
import { getTime, getTimeHumaize } from "./utils";

interface NapGachaInfo {
  ann_id: number;
  title: string;
  image: string;
  content: string;
  start_time: string;
  end_time: string;
}

interface NapProgress {
  text?: string | null;
  percent?: number | null;
}

interface NapResponse {
  progress: NapProgress;
  gacha_info: NapGachaInfo[];
}

interface AnnContentResponse {
  retcode: number;
  message: string;
  data: {
    list: {
      ann_id: number;
      title: string;
      subtitle: string;
      banner: string;
      content: string;
      lang: string;
    }[];
    pic_list: unknown[];
    total: number;
    pic_total: number;
  };
}

interface AnnListResponse {
  retcode: number;
  message: string;
  data: {
    list: {
      list: {
        ann_id: number;
        title: string;
        subtitle: string;
        banner: string;
        content: string;
        type_label: string;
        tag_label: string;
        tag_icon: string;
        login_alert: number;
        lang: string;
        start_time: string;
        end_time: string;
        type: number;
        remind: number;
        alert: number;
        tag_start_time: string;
        tag_end_time: string;
        remind_ver: number;
        has_content: boolean;
        extra_remind: number;
        tag_icon_hover: string;
      }[];
      type_id: number;
      type_label: string;
    }[];
    total: number;
    type_list: {
      id: number;
      name: string;
      mi18n_name: string;
    }[];
    alert: boolean;
    alert_id: number;
    timezone: number;
    t: string;
    pic_list: unknown[];
    pic_total: number;
    pic_type_list: unknown[];
    pic_alert: boolean;
    pic_alert_id: number;
    static_sign: string;
  };
}

async function getAnnList(): Promise<AnnListResponse> {
  const response = await fetch(
    "https://announcement-api.mihoyo.com/common/nap_cn/announcement/api/getAnnList?" +
      new URLSearchParams({
        game: "nap",
        game_biz: "nap_cn",
        lang: "zh-cn",
        bundle_id: "nap_cn",
        platform: "pc",
        region: "prod_gf_cn",
        level: "40",
        channel_id: "1",
        uid: "10000000",
      }).toString()
  );
  if (response.status !== 200) {
    throw new Error(`Fail to get ann list ${response.status}`);
  }
  if (
    response.headers.get("Content-Type")?.includes("application/json") === false
  ) {
    throw new Error(
      `Fail to get ann list ${response.headers.get("Content-Type")}`
    );
  }
  return await response.json();
}

function getVersionInfoFromAnnList(
  annList: Awaited<ReturnType<typeof getAnnList>>
):
  | {
      start_time: string;
      end_time: string;
    }
  | undefined {
  for (const lst of annList.data.list) {
    for (const i of lst.list) {
      if (
        i.title.includes("已知问题及游戏优化说明") ||
        i.subtitle.includes("版本更新说明")
      ) {
        return i;
      }
    }
  }
}

async function getAnnContent(): Promise<AnnContentResponse> {
  const response = await fetch(
    "https://announcement-api.mihoyo.com/common/nap_cn/announcement/api/getAnnContent?" +
      new URLSearchParams({
        game: "nap",
        game_biz: "nap_cn",
        lang: "zh-cn",
        bundle_id: "nap_cn",
        platform: "pc",
        region: "prod_gf_cn",
        level: "40",
        channel_id: "1",
        uid: "10000000",
      }).toString()
  );
  if (response.status !== 200) {
    throw new Error(`Fail to get ann content ${response.status}`);
  }
  if (
    response.headers.get("Content-Type")?.includes("application/json") === false
  ) {
    throw new Error(
      `Fail to get ann list ${response.headers.get("Content-Type")}`
    );
  }
  return await response.json();
}

function getGachaInfoFromAnnContent(
  annContent: Awaited<ReturnType<typeof getAnnContent>>
): {
  content: string;
  ann_id: number;
  title: string;
  image: string;
}[] {
  return annContent.data.list
    .filter((i) => i.subtitle.includes("调频"))
    .map((i) => {
      return {
        content: i.content,
        ann_id: i.ann_id,
        title: i.subtitle,
        image: i.banner,
      };
    });
}

export async function getNapInfo(): Promise<NapResponse> {
  const [annList, annContent] = await Promise.all([
    getAnnList(),
    getAnnContent(),
  ]);

  const versionInfo = getVersionInfoFromAnnList(annList);
  let progressPercent: number | null = null;
  let progressText: string | null = null;
  const gachaInfo: NapGachaInfo[] = [];

  if (versionInfo) {
    const startTime = getTime(versionInfo.start_time);
    const endTime = getTime(versionInfo.end_time);
    const currentTime = getTime();
    if (currentTime.isBetween(startTime, endTime)) {
      progressPercent = currentTime.diff(startTime) / endTime.diff(startTime);
      const durationString = getTimeHumaize(endTime);
      progressText = `${startTime.format(
        "YYYY-MM-DD HH:mm:ss"
      )} ~ ${endTime.format("YYYY-MM-DD HH:mm:ss")} （${durationString}结束）`;
    }
  }

  getGachaInfoFromAnnContent(annContent).forEach((i) => {
    let start_time: string = "";
    let end_time: string = "";
    const t =
      /(?:([0-9]+\.[0-9]版本更新后)|(\d{4}\/\d{2}\/\d{2} \d{2}:\d{2}(?::\d{2})?)).*?(\d{4}\/\d{2}\/\d{2} \d{2}:\d{2}(?::\d{2})?)/.exec(
        i.content
      );
    const groups = Array.from(t || []).slice(1) || [];
    if (groups[0] && groups[2]) {
      start_time = groups[0];
      const end = getTime(groups[2]);
      end_time = `${end.format("YYYY-MM-DD HH:mm:ss")} （${getTimeHumaize(
        end
      )}）`;
    }

    if (groups[1] && groups[2]) {
      const start = getTime(groups[1]);
      const end = getTime(groups[2]);
      start_time = `${start.format(
        "YYYY-MM-DD HH:mm:ss"
      )} （${getTimeHumaize(start)}）`;
      end_time = `${end.format("YYYY-MM-DD HH:mm:ss")} （${getTimeHumaize(
        end
      )}）`;
    }

    gachaInfo.push({
      ann_id: i.ann_id,
      title: i.title,
      image: i.image,
      content: i.content,
      start_time: start_time,
      end_time: end_time,
    });
  });

  return {
    progress: {
      percent: progressPercent,
      text: progressText,
    },
    gacha_info: gachaInfo,
  };
}
